compiler+runtime/src/cpp/jank/runtime/obj/ratio.cpp

```diff
@@ -0,0 +1,556 @@
+#include <jank/runtime/obj/ratio.hpp>
+#include <fmt/compile.h>
+#include <limits>
+
+namespace jank::runtime
+{
+
+  constexpr native_real EPSILON = std::numeric_limits<native_real>::epsilon();
+
+  obj::ratio_data::static_object(native_integer numerator, native_integer denominator)
+    : numerator{ numerator }
+    , denominator{ denominator }
+  {
+    if(denominator == 0)
+    {
+      throw std::invalid_argument("Ratio denominator cannot be zero.");
+    }
+    simplify();
+  }
+
+  object_ptr obj::ratio::create(native_integer const numerator, native_integer const denominator)
+  {
+    auto ratio_data = make_box<obj::ratio_data>(numerator, denominator);
+    if(ratio_data->denominator == 1)
+    {
+      return make_box<obj::integer>(ratio_data->numerator);
+    }
+    auto r(make_box<obj::ratio>());
+    r->data.numerator = ratio_data->numerator;
+    r->data.denominator = ratio_data->denominator;
+    return r;
+  }
+
+  native_real obj::ratio_data::to_real() const
+  {
+    return static_cast<native_real>(numerator) / denominator;
+  }
+
+  native_real obj::ratio::to_real() const
+  {
+    return data.to_real();
+  }
+
+  native_integer obj::ratio::to_integer() const
+  {
+    return data.numerator / data.denominator;
+  }
+
+  void obj::ratio_data::simplify()
+  {
+    int gcd = std::gcd(numerator, denominator);
```

Comment:
Please always use direct initialization and avoid spelling out obvious types.

```cpp
auto const gcd{ std::gcd(numerator, denominator) };
```

Reply:
done.